_posts/2020-04-18-methods-arguments-ruby-bart3.markdown

```diff
@@ -0,0 +1,257 @@
+
+Let's say we have a generic method to multiply two numbers:
+
+```ruby
+def multi(num1, num2)
+  puts num1 * num2
+end
+```
+
+Now let's imagine we want to have a method that multiplies by 3 because we are using that a lot. We can reuse that original method and only apply the number 3 using the `curry` method on the method object:
+
+```ruby
+multi3 = method(:multi).curry.call(3)
+#        ^ we use method(:mutli) to get the method object
+#          instead of calling the method
+```
+
+Now we can use that new method (it's actually a Proc) to have different numbers multipied by 3:
+
+```ruby
+multi3.call(2) # => 6
+multi3.call(5) # => 15
+
+# .() is a shorthand for .call()
+multi3.(10) # => 30
+```
+
+## Case Study 1: link_to
+
+Let's analyze Rails' `link_to` method.
+
+```ruby
+# all arguments are optional, it just prints an empty "A" tag
+def link_to(name = nil, options = nil, html_options = nil, &block)
+  # when we call this method with a block, our first argument is
+  # not the name for the link, it's actually the options, and the
+  # second argument is actually the html_options for the link
+  html_options, options, name = options, name, block if block_given?
+  
+  # it sets default options here instead of using a default value for
+  # the parameter because the previous line changes what each
+  # parameter actually is depending if there a block given
+  options ||= {}
+
+  html_options = convert_options_to_data_attributes(options, html_options)
+
+  url = url_for(options)
+  html_options["href".freeze] ||= url
+
+  # it's a wrapper over 'content_tag' method
+  content_tag("a".freeze, name || url, html_options, &block)
+end
+```
+
+## Case Study 2: pluralize
+
+Let's analyize another example: the `pluralize` helper method:
+
+```ruby
+# it requires a count and a word in singular
+
+# we have two ways to set the 'plural' of the word:
+# - using the third positional argument
+# - using the 'plural' keyword argument
+# the keyword argument uses the positional as the default value!
+
+# if we leave the plural form empty, it will use the I18n module
+# to infere the plural form, we can also include a 'locale' keyword
+# argument but it defaults to the current locale
+def pluralize(count, singular, plural_arg = nil, plural: plural_arg, locale: I18n.locale)
+  word = if (count == 1 || count.to_s =~ /^1(\.0+)?$/)
+    singular
+  else
+    # if we didn't provide a plural form, it uses the string's pluralize method
+    plural || singular.pluralize(locale)
+  end
+
+  "#{count || 0} #{word}"
+end
+```
+
+## Conclusion
+
+We have covered more ways to make our methods super flexible so that we can use the methods and customize their behavior to our needs.
```

Comment:
@arielj hm, something sounds off here. maybe:
```suggestion
We have covered more ways to make our methods super flexible so that we can use and customize their behavior to our needs.
```